lib/backend/k8s/resources/retrywrapper.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resources
+
+import (
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const (
+	maxActionRetries = 5
+)
+
+// retryWrapper implements the K8sResourceClient interface and is used to wrap
+// other clients to perform retry processing when there is an unexpected
+// resource update conflict.  This may happen when Calico data is attached to
+// a Kubernetes object that is not used solely for the Calico resource.  In this
+// case, appending the Calico data requires a two stage Get/Update to perform any
+// CRUD operation which has the potential of resulting in an update conflict
+// between the Get and the Update of the underlying resource.
+//
+// This wrapper retries the action if there is a resource update conflict but
+// there is no revision information in the KVPair - in this case we would
+// not expect an update conflict.  Actions are retried a maximum number of
+// times.
+type retryWrapper struct {
+	client K8sResourceClient
+}
+
+func (r *retryWrapper) Create(object *model.KVPair) (*model.KVPair, error) {
+	var kvp *model.KVPair
+	var err error
+	for i := 0; i < maxActionRetries; i++ {
+		if kvp, err = r.client.Create(object); err == nil {
+			// No error, exit returning the KVPair.
+			return kvp, nil
+		} else if !r.errorIndicatesRetry(object, err) {
+			return nil, err
+		}
+	}
+
+	// Excessive retries.  Return the last error.
+	log.WithField("Key", object.Key).Warning("Failed to create object: too many update conflicts.")
+	return nil, err
+}
+
+func (r *retryWrapper) Update(object *model.KVPair) (*model.KVPair, error) {
+	var kvp *model.KVPair
+	var err error
+	for i := 0; i < maxActionRetries; i++ {
+		if kvp, err = r.client.Update(object); err == nil {
```

Comment:
Yup this wrapper is wrapped around the outside of the BGPPeer Client.  So if that client filters up a "please retry this" error, this wrapper will re-invoke the Update on the BGPPeer Client (which will perform the Get/Modify/Update).